### PR TITLE
EOS-25521: [K8S:RAW] System health changes required for K8S Alert

### PR DESCRIPTION
**Pull request by Madhura-08**
# Problem Statement
- https://jts.seagate.com/browse/EOS-25521
System health changes for k8salert

# Design
systems.atlassian.net/wiki/spaces/PUB/pages/730890496/Fault+tolerance+LLD?focusedCommentId=793509919#comment-793509919)

# Coding
-  [x] Coding conventions are followed and code is consistent

# Testing 
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Review Checklist 
- [x] PR is self reviewed
- [x] JIRA number/GitHub Issue added to PR
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained
- [ ] Is there a change in filename/package/module or signature? [Y/N]: 
- [ ] If yes for above point, is a notification sent to all other cortx components? [Y/N]
- [ ] Side effects on other features (deployment/upgrade)? [Y/N]
- [ ] Dependencies on other component(s)? [Y/N]
-     If yes for above point, post link to the corresponding PR.

# Review Checklist 
- [ ] Is perfline test run and the report with and without the changes updated in the PR? [Y/N]: 

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

1. Static test was done by directly calling event_analyzerd with static alert sent:

    msg = b'{ "resource_type": "k8s:pod", "resource_name": "dummy-pod", "event_type": "online", \
              "k8s_container": "None", "pod": "None", "node": "ssc-vm-g3-rhev4-1313.colo.seagate.com", \
              "is_status": "False", "timestamp": "1636527179"}'
    

So, this EventAnalyzer will call System Health process_event().
After above execution, found following keys:
cortx/ha/v1/cortx/ha/system/cluster/12345/health:{"events": [{"event_timestamp": "1636527179", "created_timestamp": "1637306870", "status":      "online", "specific_info": null}], "action": {"modified_timestamp": "1637306870", "status": "pending"}, "attributes": {}}

cortx/ha/v1/cortx/ha/system/cluster/12345/site/1/health:{"events": [{"event_timestamp": "1636527179", "created_timestamp": "1637306870", "s     tatus": "online", "specific_info": null}], "action": {"modified_timestamp": "1637306870", "status": "pending"}, "attributes": {}}

cortx/ha/v1/cortx/ha/system/cluster/12345/site/1/rack/1/health:{"events": [{"event_timestamp": "1636527179", "created_timestamp": "16373068     70", "status": "online", "specific_info": null}], "action": {"modified_timestamp": "1637306870", "status": "pending"}, "attributes": {}}

cortx/ha/v1/cortx/ha/system/cluster/12345/site/1/rack/1/node/dummy-pod/health:{"events": [{"event_timestamp": "1636527179", "created_timest     amp": "1637306869", "status": "online", "specific_info": "specific_info"}], "action": {"modified_timestamp": "1637306869", "status": "pendi     ng"}, "attributes": {}}

Latest testing details:

New consul keys:
[root@ssc-vm-g3-rhev4-1313 cortx-ha-1]# consul kv get -http-addr=127.0.0.1:80 --recurse cortx/ha/v1                               cortx/ha/v1:
cortx/ha/v1/action/node/failed:["publish"]
cortx/ha/v1/action/node/online:["publish"]
cortx/ha/v1/events/node/failed:["hare"]
cortx/ha/v1/events/node/online:["hare"]
cortx/ha/v1/events/subscribe/hare:["node/online", "node/failed"]
cortx/ha/v1/message_type/hare:ha_event_hare

New alert:
2021-11-26 00:43:47 pod_monitor [5673]: INFO [run] Sending alert on message bus {'_resource_type': 'node', '_resource_name': 'dummy-pod-1', '_event_type': 'online', '_k8s_container': None, '_pod': None, '_node': 'ssc-vm-g3-rhev4-1313.colo.seagate.com', '_is_status': False, '_timestamp': '1637912627'}

1st POD is online:
cortx/ha/v1/cortx/ha/system/cluster/9995f539f4f770e2a3fe9e2e615c32a8/health:{"events": [{"event_timestamp": "1637918178", "created_timestamp": "1637918186", "status": "online", "specific_info": null}], "action": {"modified_timestamp": "1637918186", "status": "pending"}, "attributes": {}}
cortx/ha/v1/cortx/ha/system/cluster/9995f539f4f770e2a3fe9e2e615c32a8/site/1/health:{"events": [{"event_timestamp": "1637918178", "created_timestamp": "1637918186", "status": "online", "specific_info": null}], "action": {"modified_timestamp": "1637918186", "status": "pending"}, "attributes": {}}
cortx/ha/v1/cortx/ha/system/cluster/9995f539f4f770e2a3fe9e2e615c32a8/site/1/rack/1/health:{"events": [{"event_timestamp": "1637918178", "created_timestamp": "1637918186", "status": "online", "specific_info": null}], "action": {"modified_timestamp": "1637918186", "status": "pending"}, "attributes": {}}
cortx/ha/v1/cortx/ha/system/cluster/9995f539f4f770e2a3fe9e2e615c32a8/site/1/rack/1/node/dummy-pod-2/health:{"events": [{"event_timestamp": "1637918178", "created_timestamp": "1637918186", "status": "online", "specific_info": "specific_info"}], "action": {"modified_timestamp": "1637918186", "status": "pending"}, "attributes": {}}

2nd POD online
cortx/ha/v1/cortx/ha/system/cluster/9995f539f4f770e2a3fe9e2e615c32a8/site/1/rack/1/node/dummy-pod-2/health:{"events": [{"event_timestamp": "1637918178", "created_timestamp": "1637918186", "status": "online", "specific_info": "specific_info"}], "action": {"modified_timestamp": "1637918186", "status": "pending"}, "attributes": {}}

3rd POD online:
cortx/ha/v1/cortx/ha/system/cluster/9995f539f4f770e2a3fe9e2e615c32a8/site/1/rack/1/node/dummy-pod/health:{"events": [{"event_timestamp": "1637918417", "created_timestamp": "1637918417", "status": "online", "specific_info": "specific_info"}], "action": {"modified_timestamp": "1637918417", "status": "pending"}, "attributes": {}}

3 alerts for POD on ha_event_hare message_type:
MESSAGE:  b'{"version": "2.0", "event_type": "online", "event_id": "163791818633b3825860bb400597ac5b25c24b9ef4", "resource_type": "node", "cluster_id": "9995f539f4f770e2a3fe9e2e615c32a8", "site_id": "1", "rack_id": "1", "storageset_id": "dummy-pod-2", "node_id": "dummy-pod-2", "resource_id": "dummy-pod-2", "timestamp": "1637918178", "event_specific_info": "specific_info"}'
MESSAGE:  b'{"version": "2.0", "event_type": "online", "event_id": "163791826421aecd8f71994061acf96b2c5122c752", "resource_type": "node", "cluster_id": "9995f539f4f770e2a3fe9e2e615c32a8", "site_id": "1", "rack_id": "1", "storageset_id": "dummy-pod-1", "node_id": "dummy-pod-1", "resource_id": "dummy-pod-1", "timestamp": "1637918264", "event_specific_info": "specific_info"}'
MESSAGE:  b'{"version": "2.0", "event_type": "online", "event_id": "1637918417ab4bbaae249a4988b07a49d83250bf1d", "resource_type": "node", "cluster_id": "9995f539f4f770e2a3fe9e2e615c32a8", "site_id": "1", "rack_id": "1", "storageset_id": "dummy-pod", "node_id": "dummy-pod", "resource_id": "dummy-pod", "timestamp": "1637918417", "event_specific_info": "specific_info"}'

1 POD is down:
cortx/ha/v1/cortx/ha/system/cluster/9995f539f4f770e2a3fe9e2e615c32a8/site/1/rack/1/node/dummy-pod-1/health:{"events": [{"event_timestamp": "1637918573", "created_timestamp": "1637918573", "status": "failed", "specific_info": "specific_info"}, {"event_timestamp": "1637918264", "created_timestamp": "1637918264", "status": "online", "specific_info": "specific_info"}], "action": {"modified_timestamp": "1637918573", "status": "pending"}, "attributes": {}}
cortx/ha/v1/cortx/ha/system/cluster/9995f539f4f770e2a3fe9e2e615c32a8/health:{"events": [{"event_timestamp": "1637918573", "created_timestamp": "1637918573", "status": "degraded", "specific_info": null}, {"event_timestamp": "1637918178", "created_timestamp": "1637918186", "status": "online", "specific_info": null}], "action": {"modified_timestamp": "1637918573", "status": "pending"}, "attributes": {}}

HARE event:
{"version": "2.0", "event_type": "failed", "event_id": "16379185736160c5a70d8c43e4b47bc10e115a81be", "resource_type": "node", "cluster_id": "9995f539f4f770e2a3fe9e2e615c32a8", "site_id": "1", "rack_id": "1", "storageset_id": "dummy-pod-1", "node_id": "dummy-pod-1", "resource_id": "dummy-pod-1", "timestamp": "1637918573", "event_specific_info": "specific_info"}

2nd POD is down:
cortx/ha/v1/cortx/ha/system/cluster/9995f539f4f770e2a3fe9e2e615c32a8/site/1/rack/1/node/dummy-pod-2/health:{"events": [{"event_timestamp": "1637918737", "created_timestamp": "1637918737", "status": "failed", "specific_info": "specific_info"}, {"event_timestamp": "1637918178", "created_timestamp": "1637918186", "status": "online", "specific_info": "specific_info"}], "action": {"modified_timestamp": "1637918737", "status": "pending"}, "attributes": {}}
cortx/ha/v1/cortx/ha/system/cluster/9995f539f4f770e2a3fe9e2e615c32a8/health:{"events": [{"event_timestamp": "1637918737", "created_timestamp": "1637918737", "status": "failed", "specific_info": null}, {"event_timestamp": "1637918573", "created_timestamp": "1637918573", "status": "degraded", "specific_info": null}], "action": {"modified_timestamp": "1637918737", "status": "pending"}, "attributes": {}}

HARE event:
{"version": "2.0", "event_type": "failed", "event_id": "1637918737ca351178affc4d9097a9ba07268906ad", "resource_type": "node", "cluster_id": "9995f539f4f770e2a3fe9e2e615c32a8", "site_id": "1", "rack_id": "1", "storageset_id": "dummy-pod-2", "node_id": "dummy-pod-2", "resource_id": "dummy-pod-2", "timestamp": "1637918737", "event_specific_info": "specific_info"}

### CODE REVIEW

**Comment by ArchanaLimaye** [comment]
Please complete the PR template

File: ha/core/event_analyzer/parser/parser.py [hunk 1]
```diff
@@ -145,3 +151,60 @@ def parse_event(self, msg: str) -> HealthEvent:
...
+            msg (str): Msg
+        """
+        try:
+            cluster_resource_alert = json.loads(msg)
+            cluster_id = Conf.get(const.HA_GLOBAL_INDEX, f"COMMON_CONFIG{_DELIM}cluster_id")
+            site_id = Conf.get(const.HA_GLOBAL_INDEX, f"COMMON_CONFIG{_DELIM}site_id")
+            rack_id = Conf.get(const.HA_GLOBAL_INDEX, f"COMMON_CONFIG{_DELIM}rack_id")
+            timestamp = str(int(time.time()))
+            event_id = timestamp + str(uuid.uuid4().hex)
+            pod_id = cluster_resource_alert["_resource_name"]
```
Comment (Madhura-08) [inline ha/core/event_analyzer/parser/parser.py:182]: rename this pod_id to node_id

File: ha/k8s_setup/ha_setup.py [hunk 2]
```diff
@@ -89,6 +89,12 @@ def get_command(desc: str, argv: dict):
         args = parser.parse_args(argv)
         return args.command(args)
 
+    def get_machine_id(self):
```
Comment (Madhura-08) [inline ha/k8s_setup/ha_setup.py:92]: Verify whether this will actually give cluster_id

**Review by shaileshsvc — APPROVED** [review]
Core logic in context of DoD for this Jira is reviewed and approved.
[Renames of pod_id to mode_id, resource_type assignment and codacy issue resolution will be covered in later reviews by Archana. Filter logic, site, rack and cluster-id logic may need changes later in integration story once definitive entries in cluster.conf are available for the same.]

File: ha/core/event_analyzer/parser/parser.py (hunk 1 above)
Comment (Madhura-08) [reply ha/core/event_analyzer/parser/parser.py]: Done

File: ha/k8s_setup/ha_setup.py (hunk 2 above)
Comment (Madhura-08) [reply ha/k8s_setup/ha_setup.py:92]: Already verified in the meeting

File: ha/core/event_analyzer/filter/filter.py [hunk 3]
```diff
@@ -174,7 +174,6 @@ def __init__(self):
         Init method
         """
         super(ClusterResourceFilter, self).__init__()
-        ConfigManager.init("event_analyzer")
```
Comment (ArchanaLimaye) [inline ha/core/event_analyzer/filter/filter.py:177]: Why is this removed ?

File: ha/core/event_analyzer/parser/parser.py [hunk 4]
```diff
@@ -145,3 +152,60 @@ def parse_event(self, msg: str) -> HealthEvent:
...
+        Args:
+            msg (str): Msg
+        """
+        try:
+            message = json.dumps(ast.literal_eval(msg))
+            cluster_resource_alert = json.loads(message)
+            cluster_id = Conf.get(const.HA_GLOBAL_INDEX, f"COMMON_CONFIG{_DELIM}cluster_id")
+            site_id = Conf.get(const.HA_GLOBAL_INDEX, f"COMMON_CONFIG{_DELIM}site_id")
+            rack_id = Conf.get(const.HA_GLOBAL_INDEX, f"COMMON_CONFIG{_DELIM}rack_id")
+            timestamp = str(int(time.time()))
```
Comment (ArchanaLimaye) [inline ha/core/event_analyzer/parser/parser.py:182]: This can be deleted

File: ha/core/system_health/system_health.py [hunk 5]
```diff
@@ -304,7 +304,7 @@ def publish_event(self, healthevent: HealthEvent, healthvalue: str= ""):
         """
         healthevent.event_type = json.loads(healthvalue).get("events")[0]["status"]
         node_id = healthevent.node_id
-        healthevent.node_id = ConfigManager.get_node_name(str(node_id))
+        # healthevent.node_id = ConfigManager.get_node_name(str(node_id))
```
Comment (ArchanaLimaye) [inline ha/core/system_health/system_health.py:307]: can delete this

File: ha/fault_tolerance/fault_tolerance_driver.py [hunk 6]
```diff
@@ -60,24 +58,21 @@ def _get_consumer(self) -> MessageBusConsumer:
...
         """Contineously polls for message bus for k8s_event message type"""
         try:
             self._consumer.start()
             while True:
-                # Get alert from message. Analyze changes
-                # with the help of event analyzer filter and publish to message bus
-                # if required
+                # Get alert condition from ALertGenerator. Analyze changes
+                # with the help of event analyzer and notify if required
                 Log.info('Ready to analyze faults in the system')
```
Comment (ArchanaLimaye) [inline ha/fault_tolerance/fault_tolerance_driver.py:76]: this log is not needed inside the while loop

File: ha/k8s_setup/ha_setup.py [hunk 7]
```diff
@@ -199,9 +219,21 @@ def process(self):
             with open(const.HA_CONFIG_FILE, 'w+') as conf_file:
                 yaml.dump(conf_file_dict, conf_file, default_flow_style=False)
 
+            Cmd.copy_file(const.SOURCE_HEALTH_HIERARCHY_FILE, const.HEALTH_HIERARCHY_FILE)
             # First populate the ha.conf and then do init. Because, in the init, this file will
             # be stored in the confstore as key values
             ConfigManager.init("ha_setup")
+
+            machine_id = self.get_machine_id()
```
Comment (ArchanaLimaye) [inline ha/k8s_setup/ha_setup.py:227]: Please add a comment why this machine Id is being fetched here

File: ha/k8s_setup/ha_setup.py [hunk 8]
```diff
@@ -177,6 +195,7 @@ def process(self):
             # Dummy value fetched for now. This will be replaced by the key/path for the pod label onces that is avilable in confstore
             # Ref ticket EOS-25694
             data_pod_label = Conf.get(self._index, f'cortx{_DELIM}common{_DELIM}product_release')
+            # cluster_id = Conf.get(self._index, f'node{_DELIM}{machine_id}{_DELIM}cluster_id')
```
Comment (ArchanaLimaye) [inline ha/k8s_setup/ha_setup.py:198]: If we are getting cluster Id here remove the code for the same from process function, otherwise remove this (commented)  line

File: ha/k8s_setup/ha_setup.py [hunk 9]
```diff
@@ -199,9 +219,21 @@ def process(self):
...
             # be stored in the confstore as key values
             ConfigManager.init("ha_setup")
+
+            machine_id = self.get_machine_id()
+            cluster_id = Conf.get(self._index, f'node{_DELIM}{machine_id}{_DELIM}cluster_id')
+            # site_id = Conf.get(self._index, f'node{_DELIM}{machine_id}{_DELIM}site_id')
+            site_id = '1'
+            # rack_id = Conf.get(self._index, f'node{_DELIM}{machine_id}{_DELIM}rack_id')
+            rack_id = '1'
+            conf_file_dict.update({'COMMON_CONFIG': {'cluster_id': cluster_id, 'rack_id': rack_id, 'site_id': site_id}})
```
Comment (ArchanaLimaye) [inline ha/k8s_setup/ha_setup.py:233]: How will this get updated into consul ?

File: ha/core/event_analyzer/filter/filter.py (hunk 3 above)
Comment (Madhura-08) [reply ha/core/event_analyzer/filter/filter.py:177]: This is not required here now. It will be called from EventAnalyzer class where init will happen

File: ha/core/event_analyzer/parser/parser.py (hunk 4 above)
Comment (Madhura-08) [reply ha/core/event_analyzer/parser/parser.py:182]: We can not remove this. Its needed for event_id creation

File: ha/core/system_health/system_health.py (hunk 5 above)
Comment (Madhura-08) [reply ha/core/system_health/system_health.py]: Done.

File: ha/fault_tolerance/fault_tolerance_driver.py (hunk 6 above)
Comment (Madhura-08) [reply ha/fault_tolerance/fault_tolerance_driver.py]: Done

File: ha/k8s_setup/ha_setup.py (hunk 7 above)
Comment (Madhura-08) [reply ha/k8s_setup/ha_setup.py:231]: Done

File: ha/k8s_setup/ha_setup.py (hunk 8 above)
Comment (Madhura-08) [reply ha/k8s_setup/ha_setup.py]: Removed it from here.

File: ha/k8s_setup/ha_setup.py (hunk 9 above)
Comment (Madhura-08) [reply ha/k8s_setup/ha_setup.py:237]: Just below lines have code to dump this data in yaml format to ha.conf

Comment (ArchanaLimaye) [reply ha/k8s_setup/ha_setup.py:237]: yes, but the configManager init was going to pickup the ha.conf data and put it in confstore right ? where will that happen ?

Comment (Madhura-08) [reply ha/k8s_setup/ha_setup.py:237]: As discussed, it will be available in the confstore after starting one of the service. But just added TODO here to make sure everything is in place

File: ha/fault_tolerance/fault_tolerance_driver.py [hunk 10]
```diff
@@ -60,25 +58,21 @@ def _get_consumer(self) -> MessageBusConsumer:
...
         """Callback method for MessageConsumer"""
-        Log.debug(f'Received the message from message bus: {message}')
-        result = False
-        # EventAnalyzer class instantiation is expected.
-        # This is not required. This will be called from EventAnalyzer class.
-        # But as its not available yet and to check newly added filter functionality,
-        # temporarily added this
-        result = self._cluster_resource_filter.filter_event(message.decode('utf-8'))
-        Log.info(f'Alert required: {result}')
+        Log.info(f'Received the message from message bus: {message}')
```
Comment (ArchanaLimaye) [inline ha/fault_tolerance/fault_tolerance_driver.py:61]: should be log.debug  , otherwise it will flood the log

File: ha/fault_tolerance/fault_tolerance_driver.py [hunk 11]
```diff
@@ -60,25 +58,21 @@ def _get_consumer(self) -> MessageBusConsumer:
...
+        try:
+            EventAnalyzer(message.decode('utf-8'))
+        except Exception as err:
+            Log.error(f'Failed to analyze the event: {err}')
+            return CONSUMER_STATUS.FAILED
```
Comment (ArchanaLimaye) [inline ha/fault_tolerance/fault_tolerance_driver.py:66]: Confrim that this return status is true. If the message itself was corrupted we should not ask for the same message to be read again otherwise will be stuck in infinite loop

File: ha/core/event_analyzer/parser/parser.py [hunk 12]
```diff
@@ -145,3 +152,60 @@ def parse_event(self, msg: str) -> HealthEvent:
...
+                EVENT_ATTRIBUTES.HOST_ID : node_id,
+                EVENT_ATTRIBUTES.RESOURCE_TYPE : resource_type,
+                EVENT_ATTRIBUTES.TIMESTAMP : timestamp,
+                EVENT_ATTRIBUTES.RESOURCE_ID : node_id,
+                EVENT_ATTRIBUTES.SPECIFIC_INFO : "specific_info"
+            }
+
+            Log.debug(f"Parsed {event} schema")
+            health_event = HealthEvent.dict_to_object(event)
+            Log.info(f"Event {event[EVENT_ATTRIBUTES.EVENT_ID]} is parsed and converted to object.")
```
Comment (ArchanaLimaye) [inline ha/core/event_analyzer/parser/parser.py:207]: This should be log.debug so as not to flood the log

**Review by ArchanaLimaye — APPROVED** [review]

File: ha/fault_tolerance/fault_tolerance_driver.py (hunk 10 above)
Comment (Madhura-08) [reply ha/fault_tolerance/fault_tolerance_driver.py:61]: Done

File: ha/fault_tolerance/fault_tolerance_driver.py (hunk 11 above)
Comment (Madhura-08) [reply ha/fault_tolerance/fault_tolerance_driver.py:66]: Right. But as message is coming from message bus. We can be sure that message json structure is proper and not corrupted. So, means there can be some other issue and retry can solve that. I thought of this way.

File: ha/core/event_analyzer/parser/parser.py (hunk 12 above)
Comment (Madhura-08) [reply ha/core/event_analyzer/parser/parser.py:207]: Done

File: ha/fault_tolerance/fault_tolerance_driver.py (hunk 6 above)
Comment (ArchanaLimaye) [reply ha/fault_tolerance/fault_tolerance_driver.py]: Thanks. You can add a log before the while loop starts , to indicate the poll is started

File: ha/fault_tolerance/fault_tolerance_driver.py (hunk 11 above)
Comment (ArchanaLimaye) [reply ha/fault_tolerance/fault_tolerance_driver.py:66]: Yes, however this return value will cause message retry even if there is some exception received in system health path (which is not related to the incoming message) 
Let us handle it separately , I am closing this comment for now

File: ha/fault_tolerance/fault_tolerance_driver.py (hunk 6 above)
Comment (ArchanaLimaye) [reply ha/fault_tolerance/fault_tolerance_driver.py]: Closing this comment, we can add the log in a separate PR
